dev/encoder/quadrature_encoder.cpp

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) 2023 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <stdexcept>
+#include "pico/stdlib.h"
+#include "hardware/pio.h"
+#include "hardware/timer.h"
+
+#include "quadrature_encoder.pio.h"
+
+//
+// ---- quadrature encoder interface example
+//
+// the PIO program reads phase A/B of a quadrature encoder and increments or
+// decrements an internal counter to keep the current absolute step count
+// updated. At any point, the main code can query the current count by using
+// the quadrature_encoder_*_count functions. The counter is kept in a full
+// 32 bit register that just wraps around. Two's complement arithmetic means
+// that it can be interpreted as a 32-bit signed or unsigned value, and it will
+// work anyway.
+//
+// As an example, a two wheel robot being controlled at 100Hz, can use two
+// state machines to read the two encoders and in the main control loop it can
+// simply ask for the current encoder counts to get the absolute step count. It
+// can also subtract the values from the last sample to check how many steps
+// each wheel as done since the last sample period.
+//
+// One advantage of this approach is that it requires zero CPU time to keep the
+// encoder count updated and because of that it supports very high step rates.
+//
+
+// 374 pulses per revolution (from the description at https://www.dfrobot.com/product-1462.html)
+// 4x encoding bc quadrature (?)
+const float ROT_PER_TICK = 1.0 / (4 * 374.0);
+// TODO: update pulley ratio
+const float PULLEY_RATIO = 0.3185 / 1.528;
+const float DEG_PER_ROT = 360.0;
+
+class Encoder
+{
+public:
+    // Create an encoder. Reccommended NOT to use this class, use EncoderFactory::createEncoder()
+    // @param pinA the pin that encoder A channel is connected to, the B channel should connect to the next pin
+    // @param sm the state machine to keep track of the encoder, 0-3
+    // @param which pio
+    Encoder(uint pinA, uint sm, PIO pio, float ratio, bool addProgram = true)
+    {
+        this->pio = pio;
+        this->sm = sm;
+        this->ratio = ratio;
+
+        uint offset = 0;
+
+        // we don't really need to keep the offset, as this program must be loaded
+        // at offset 0
+        if (addProgram)
+            uint offset = pio_add_program(pio, &quadrature_encoder_program);
+
+        quadrature_encoder_program_init(pio, sm, offset, pinA, 0);
+    }
+
+    // updates the pos and velocity, call periodically.
+    // @param delta_time the time, in miliseconds, since last calling update
+    void update(int delta_time)
+    {
+        pos = quadrature_encoder_get_count(pio, sm) * ratio * DEG_PER_ROT;
+        velocity = ((prev_pos - pos) / delta_time) * 1000;
+        prev_pos = pos;
+    }
+
+    // get position of wheel in rotations. resets on init.
+    // update() must be called preiodically for this to be accurate
+    float get_pos()
```

Comment:
"get position of wheel in ~~rotations~~"
instead should be
"get position of wheel in degrees, multiplied by ratio in constructor" 
or similiar?

Reply:
see also getVelocity()

Reply:
see also createEncoder in the other class for same issue

Reply:
wait nevermind it's in ticks, so maybe add ROT_PER_TICK so that the motor is actually in degrees by default? It's kinda weird you have to pass the ticks per rotation in the constructor, when the motor should know that, but not the degrees? either do both or do neither. my preference would be both.

Reply:
that makes sense yeah, i changed those 2 comments and I'll change the constants so the constructor is literally just asking for the ratio between desired output & encoder ticks